### PR TITLE
provide ui for choosing best match of a record

### PR DESCRIPTION
**Pull request by Dinika**
Needs [#3150](https://github.com/inspirehep/inspire-next/pull/3150) on inspire-next for styling the UI

On the search results
screenshot from 2018-02-02 10-28-09

On detailed record page
screenshot from 2018-02-02 10-28-27

Signed-off-by: Dinika Saxena <dinika.saxena@cern.ch>

### CODE REVIEW

**Comment by Dinika** [comment]
The test failed needs to be fixed since it fails on the 26th of every month. PR #49 fixes it.

File: src/inspirehep-holdingpen-js/templates/inspirehep-holdingpen-js/results/results.html [hunk 1]
```diff
@@ -96,9 +96,38 @@
                 </span>
               </p>
             </div>
+            
+            <div ng-if="record._source._extra_data.matches.fuzzy.length > 1 
+            && record._source._workflow.data_type === 'hep' && record._source._workflow.status === 'HALTED'"
+            class="matches-container">
+                <div ng-if="vm.record_matched.length < 1" >
```
Comment (drjova) [inline src/inspirehep-holdingpen-js/templates/inspirehep-holdingpen-js/results/results.html]: IMHO it will prevent other matches, ``vm.record_matched`` is a controller's variable and if a users click a match all the others will disappear

**Comment by Dinika** [comment]
@drjova I've updated the PR to correct the way I was displaying the success message ('Record was matched') by using a control variable `recordMatched`.

File: test/unit/holdingpen/holdingpen.service.spec.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+  it('should post data to correct url when best match selected', function(){
+      var workflowId = 2,
+          match = 123456,
+          showPossibleMatches = [],
+          data = JSON.stringify({
+                'match_recid': match 
+              });
+           
+      HoldingPenRecordService.onBestMatchSelected(vm, workflowId, match, showPossibleMatches);
```
Comment (harunurhan) [inline test/unit/holdingpen/holdingpen.service.spec.js]: this does other things like `showPossibleMatches.push(1)`, even though I don't get the purpose, it should be included in the test.

File: test/unit/holdingpen/holdingpen.module.spec.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+ * 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+ *
+ * In applying this license, CERN does not
+ * waive the privileges and immunities granted to it by virtue of its status
+ * as an Intergovernmental Organization or submit itself to any jurisdiction.
+ */
+
+'use strict';
+
+describe('Unit: testing holdingpen dependencies', function() {
```
Comment (harunurhan) [inline test/unit/holdingpen/holdingpen.module.spec.js:26]: you could separate this another commit since it seems irrelevant to the others.

File: src/inspirehep-holdingpen-js/holdingpen/holdingpen.services.js [hunk 4]
```diff
@@ -83,6 +83,18 @@
               });
             },
 
+            onBestMatchSelected: function(vm, workflowId, match, showPossibleMatches) {
+              var data = JSON.stringify({
+                'match_recid': match 
+              });
+              
+              $http.post('/api/holdingpen/' + workflowId + '/action/resolve', data).then(function() {
+                showPossibleMatches.push(1);
```
Comment (harunurhan) [inline src/inspirehep-holdingpen-js/holdingpen/holdingpen.services.js]: It is almost always bad to have output parameters, can't we have this on `vm` if it is needed on UI?

Comment (harunurhan) [reply src/inspirehep-holdingpen-js/holdingpen/holdingpen.services.js]: In fact having those as arrays so that you carry reference and push some dummy element to have UI state, is confusing and too hackish even perhaps for this codebase.

**Review by harunurhan — CHANGES_REQUESTED** [review]

**Review by harunurhan — APPROVED** [review]

File: src/inspirehep-holdingpen-js/holdingpen/holdingpen.controllers.js [hunk 5]
```diff
@@ -78,6 +81,23 @@
     function setDecision(id, decision) {
       HoldingPenRecordService.setBatchDecision($scope.vm.invenioSearchResults.hits.hits, [+id], decision);
     }
+
+    function onBestMatchSelected(workflowId){
+      HoldingPenRecordService.setMatchDecision(workflowId, $scope.selectedBestMatchesById[workflowId])
+        .then(function () {
+          $scope.matched = true;
```
Comment (drjova) [inline src/inspirehep-holdingpen-js/holdingpen/holdingpen.controllers.js]: this will affect the whole result list, it's better to handle it per ``workflowId ``

File: src/inspirehep-holdingpen-js/templates/inspirehep-holdingpen-js/details/details_decision_hep.html [hunk 6]
```diff
@@ -46,6 +46,35 @@
   <p>No further actions available</p>
 </div>
 
+<div ng-if="vm.record._extra_data.matches.fuzzy && vm.record._workflow.data_type === 'hep' 
+            && vm.record._workflow.status === 'HALTED' && vm.record._extra_data._action == 'match_approval'"
+            class="matches-container">
+    <div ng-if="!matchDecisionMade">
+        <p class="sub-heading">Possible Matches</p>
+          <div class="matchingRecords" ng-repeat="match in vm.record._extra_data.matches.fuzzy">
+            <input id="{{match}}" ng-model="vm.selectedBestMatch" name="possible-matches" type="radio" value="{{match}}">
```
Comment (drjova) [inline src/inspirehep-holdingpen-js/templates/inspirehep-holdingpen-js/details/details_decision_hep.html]: actually is required to ``pre-select`` the first one

File: src/inspirehep-holdingpen-js/holdingpen/holdingpen.controllers.js [hunk 7]
```diff
@@ -78,6 +83,23 @@
...
+          $scope.matchedById[workflowId] = true;
+          $scope.matchDecisionMadeById[workflowId] = true;
+        });
+    }
+
+    function onNoMatchSelected(workflowId) {
+      HoldingPenRecordService.setMatchDecision(workflowId, null)
+      .then(function () {
+          $scope.matchedById[workflowId] = false;
+          $scope.matchDecisionMadeById[workflowId] = true;
```
Comment (drjova) [inline src/inspirehep-holdingpen-js/holdingpen/holdingpen.controllers.js:99]: shouldn't be ``false``?

Comment (harunurhan) [reply src/inspirehep-holdingpen-js/holdingpen/holdingpen.controllers.js:99]: nope, this is when they decide none of them matches.

**Review by drjova — APPROVED** [review]
